### PR TITLE
[Feature] Multiple eos countdown in query cache

### PR DESCRIPTION
**Pull request by satanson**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

how to work: reference to https://starrocks.feishu.cn/docx/MfIvd5C4EoJjhFx70JTcovgJnif

### CODE REVIEW

File: be/src/exec/pipeline/scan/morsel.h [hunk 1]
```diff
@@ -223,19 +235,76 @@ class FixedMorselQueue final : public MorselQueue {
...
+        if (_ticket_checker == nullptr) {
+            return default_value;
+        } else {
+            DCHECK(0 <= _tablet_idx && _tablet_idx < _tablets.size());
+            auto tablet_id = _tablets[_tablet_idx]->tablet_id();
+            return func(_ticket_checker, tablet_id);
+        }
+    }
+
+    int _inc_num_splits_unlocked() {
```
Comment (stdpain) [inline be/src/exec/pipeline/scan/morsel.h:262]: why _inc_num_splits_unlocked has the suffix _unlocked but _emit_eos_morsel and _has_emitted_eos_morsel without?

**Comment by wanpengfei-git** [comment]
BE Build Success

File: be/src/exec/pipeline/scan/eos_chunk_source.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+#pragma once
+
+#include "exec/pipeline/scan/chunk_source.h"
+#include "exec/pipeline/scan/morsel.h"
+#include "exec/pipeline/scan/olap_scan_context.h"
+namespace starrocks::pipeline {
+class EOSChunkSource : public ChunkSource {
```
Comment (imay) [inline be/src/exec/pipeline/scan/eos_chunk_source.h:9]: Please add some comments for this class.
Why is this class needed?

File: be/src/exec/pipeline/scan/eos_chunk_source.cpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#include "column/chunk.h"
+#include "exec/workgroup/work_group.h"
+namespace starrocks::pipeline {
+
+Status EOSChunkSource::_read_chunk(starrocks::RuntimeState* state, vectorized::ChunkPtr* chunk) {
+    DCHECK(chunk != nullptr);
+    auto chk = std::make_shared<vectorized::Chunk>();
+    auto [tablet_id, _] = _morsel->get_lane_owner_and_version();
+    chk->owner_info().set_owner_id(tablet_id, true);
+    *chunk = std::move(chk);
```
Comment (imay) [inline be/src/exec/pipeline/scan/eos_chunk_source.cpp:14]: Is this chunk necessary when this function returns EOF?

File: be/src/exec/query_cache/ticket_checker.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+#include <unordered_map>
+
+#include "gutil/macros.h"
+namespace starrocks::query_cache {
+
+class TicketChecker;
+using TicketCheckerRawPtr = TicketChecker*;
+using TicketCheckerPtr = std::shared_ptr<TicketChecker>;
+using TicketIdType = int64_t;
+class TicketChecker {
```
Comment (imay) [inline be/src/exec/query_cache/ticket_checker.h:16]: Please explain how to use it and what it does/

File: be/src/exec/query_cache/ticket_checker.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    static constexpr int64_t ALL_READY_BIT = 1L << 63;
+    static constexpr int64_t ENTER_COUNT_BITS = (1L << 30) - 1;
+    static constexpr int LEAVE_COUNT_SHIFT = 30;
+    static constexpr int64_t LEAVE_COUNT_BITS = (1L << 30) << LEAVE_COUNT_SHIFT;
```
Comment (imay) [inline be/src/exec/query_cache/ticket_checker.h:20]: These are not very clear.

**Comment by imay** [comment]
You need to explain what problem you want to fix in the PR and how you will fix it.
Actually, I can know nothing about the context. Please let reviewers on the same page as you.

File: be/src/exec/pipeline/scan/morsel.h (hunk 1 above)
Comment (satanson) [reply be/src/exec/pipeline/scan/morsel.h:262]: unlocked suffix is meanless, so it is removed.

**Comment by satanson** [comment]
> You need to explain what problem you want to fix in the PR and how you will fix it. Actually, I can know nothing about the context. Please let reviewers on the same page as you.

@imay 
add document to explain how to work

**Comment by satanson** [comment]
run starrocks_clang-format

**Comment by satanson** [comment]
run  starrocks_be_unittest

**Review by stdpain — APPROVED** [review]

File: be/src/exec/query_cache/ticket_checker.h [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+// to the leave_count, if that is true, we know ScanOperator receives all of the EOS chunks of
+// SplitMorsels, that means the original ScanMorsel has been processed completely and a EOS chunk
+// can emit from ScanOperator to its successor MultilaneOperator.
+
+class TicketChecker;
+using TicketCheckerRawPtr = TicketChecker*;
+using TicketCheckerPtr = std::shared_ptr<TicketChecker>;
+using TicketIdType = int64_t;
+
+class TicketChecker {
```
Comment (kangkaisen) [inline be/src/exec/query_cache/ticket_checker.h:27]: such as MorselEos name is better?

File: be/src/exec/pipeline/scan/scan_operator.cpp [hunk 7]
```diff
@@ -202,13 +202,23 @@ StatusOr<vectorized::ChunkPtr> ScanOperator::pull_chunk(RuntimeState* state) {
...
     return res;
 }
 
+std::tuple<int64_t, bool> ScanOperator::_should_emit_eos(const ChunkPtr& chunk) {
+    auto tablet_id = chunk->owner_info().owner_id();
+    auto is_last_chunk = chunk->owner_info().is_last_chunk();
+    if (is_last_chunk && _ticket_checker != nullptr) {
+        is_last_chunk = _ticket_checker->leave(tablet_id);
+    }
+    return {tablet_id, is_last_chunk};
```
Comment (kangkaisen) [inline be/src/exec/pipeline/scan/scan_operator.cpp:219]: could only return is_last_chunk and get tablet id from res?

File: be/src/exec/pipeline/pipeline_driver.cpp [hunk 8]
```diff
@@ -56,8 +57,20 @@ Status PipelineDriver::prepare(RuntimeState* runtime_state) {
 
     DCHECK(_state == DriverState::NOT_READY);
 
-    source_operator()->add_morsel_queue(_morsel_queue);
+    auto* source_op = source_operator();
+    // attach ticket_checker to both ScanOperator and SplitMorselQueue
+    auto should_attach_ticket_checker = (dynamic_cast<ScanOperator*>(source_op) != nullptr) &&
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver.cpp:62]: do we support all scan operators currently? did you test it?

File: be/src/exec/pipeline/scan/morsel.h [hunk 9]
```diff
@@ -223,19 +226,49 @@ class FixedMorselQueue final : public MorselQueue {
...
+
+    const Morsels _morsels;
+    // The number of the morsels before split them to pieces.
+    const size_t _num_original_morsels;
+    const int64_t _degree_of_parallelism;
+    // The minimum number of rows picked up from a segment at one time.
+    const int64_t _splitted_scan_rows;
+
+    std::vector<TabletSharedPtr> _tablets;
+    std::atomic<size_t> _tablet_idx = 0;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/scan/morsel.h:260]: why only _tablet_idx is std::atomic?

File: be/src/exec/pipeline/pipeline_driver.cpp (hunk 8 above)
Comment (satanson) [reply be/src/exec/pipeline/pipeline_driver.cpp:62]: only OlapScanOperator. only OlapScanOperator can turn on query cache

File: be/src/exec/pipeline/scan/morsel.h (hunk 9 above)
Comment (satanson) [reply be/src/exec/pipeline/scan/morsel.h:260]: it is a field moved from PhysicalSplitMorselQueue, it is contended by multiple threads.

File: be/src/exec/query_cache/ticket_checker.h (hunk 6 above)
Comment (satanson) [reply be/src/exec/query_cache/ticket_checker.h:27]: TicketChecker is inspired by team admission tickets,  when all of the team members gather together, the show there team ticket, the ticket checker permit them enter, after they leaves, show the team ticket again ant the ticket is collected and invalid.

File: be/src/exec/pipeline/scan/scan_operator.cpp (hunk 7 above)
Comment (satanson) [reply be/src/exec/pipeline/scan/scan_operator.cpp:219]: it is some tedious, and there is no performance issue, the compiler when find this and use register to save these values.

**Comment by satanson** [comment]
run starrocks_admit_test

**Review by murphyatwork — APPROVED** [review]

**Pull request merged**